### ISSUE TITLE
Black borders after conversion

### ISSUE THREAD

**Issue by steff517** [bash0/cewe2pdf#55]
After conversion, many images have black borders right or bottom and right like so:

![image](https://user-images.githubusercontent.com/1554347/100521848-80319f80-31a6-11eb-87c0-9b3cd5d10bbb.png)

Source code of this page is:
` <page designStyleID="125319" pagenr="42" rotation="0" type="normalpage">
        <bundlesize height="1480" width="3800"/>
        <designElementIDs background="6078" layout="125027" styles_v2="125319"/>
        <background alignment="1" designElementId="6078" rotation="0" type="1"/>
        <background designElementId="125027" layout="1" templatename="schema_landscape_053_y,normal"/>
        <area areatype="imagearea">
            <position height="1321.17" left="726.222" rotation="0" top="79.4146" width="1096.37" zposition="138"/>
            <designElementIDs passepartout="74124"/>
            <decoration>
                <shadow shadowAngle="149" shadowBlurNew="6" shadowDistance="17.98234921810608" shadowEnabled="1" shadowIntensity="41" shadowWidthInMM="0.0070370370370370378"/>
            </decoration>
            <image filename="safecontainer:/2020-09-17 14-14-07 - (DSC04007).JPG" passepartoutDesignElementId="74124" useABK="1">
                <cutout left="0" scale="0.26529" top="-163.078"/>
                <ClipartConfiguration applySpotColor="0"/>
            </image>
        </area>
        <area areatype="imagearea">
            <position height="632.429" left="76.7037" rotation="0" top="79.4146" width="593.926" zposition="139"/>
            <designElementIDs passepartout="74126"/>
            <decoration>
                <shadow shadowAngle="149" shadowBlurNew="6" shadowDistance="17.98234921810608" shadowEnabled="1" shadowIntensity="41" shadowWidthInMM="0.0070370370370370378"/>
            </decoration>
            <image filename="safecontainer:/2020-09-17 14-14-59 - (DSC04011).JPG" passepartoutDesignElementId="74126" useABK="1">
                <cutout left="0" scale="0.142182" top="-120.571"/>
                <ClipartConfiguration applySpotColor="0"/>
            </image>
        </area>
        <area areatype="imagearea">
            <position height="542.542" left="1965.44" rotation="0" top="79.4146" width="1006.3" zposition="140"/>
            <designElementIDs passepartout="74128"/>
            <decoration>
                <shadow shadowAngle="149" shadowBlurNew="6" shadowDistance="17.98234921810608" shadowEnabled="1" shadowIntensity="41" shadowWidthInMM="0.0070370370370370378"/>
            </decoration>
            <image filename="safecontainer:/2020-09-18 09-38-42 IMG_20200918_093841.jpg" passepartoutDesignElementId="74128" useABK="1">
                <cutout left="-8.65173e-14" scale="0.190254" top="-71.0177"/>
                <ClipartConfiguration applySpotColor="0"/>
            </image>
        </area>
        <area areatype="imagearea">
            <position height="542.542" left="3028.04" rotation="0" top="91.242" width="523.556" zposition="141"/>
            <designElementIDs passepartout="74124"/>
            <decoration>
                <shadow shadowAngle="149" shadowBlurNew="6" shadowDistance="17.98234921810608" shadowEnabled="1" shadowIntensity="41" shadowWidthInMM="0.0070370370370370378"/>
            </decoration>
            <image filename="safecontainer:/2020-09-18 09-42-52 IMG_20200918_094251_1.jpg" passepartoutDesignElementId="74124" useABK="1">
                <cutout left="-275.674" scale="0.560945" top="-44.9438"/>
                <ClipartConfiguration applySpotColor="0"/>
            </image>
        </area>
        <area areatype="imagearea">
            <position height="632.429" left="2149.11" rotation="0" top="768.878" width="523.556" zposition="142"/>
            <designElementIDs passepartout="74126"/>
            <decoration>
                <shadow shadowAngle="149" shadowBlurNew="6" shadowDistance="17.98234921810608" shadowEnabled="1" shadowIntensity="41" shadowWidthInMM="0.0070370370370370378"/>
            </decoration>
            <image filename="safecontainer:/2020-10-20 16-09-12 (2020-10-20 16-09-12 IMG_20201020_160911).jpg" passepartoutDesignElementId="74126" useABK="1">
                <cutout left="0" scale="0.130558" top="-18.7924"/>
                <ClipartConfiguration applySpotColor="0"/>
            </image>
        </area>
        <area areatype="imagearea">
            <position height="632.429" left="2728.26" rotation="0" top="768.878" width="1006.3" zposition="143"/>
            <designElementIDs passepartout="74128"/>
            <decoration>
                <shadow shadowAngle="149" shadowBlurNew="6" shadowDistance="17.98234921810608" shadowEnabled="1" shadowIntensity="41" shadowWidthInMM="0.0070370370370370378"/>
            </decoration>
            <image filename="safecontainer:/2020-10-20 16-10-07 (2020-10-20 16-10-07 IMG_20201020_161006).jpg" passepartoutDesignElementId="74128" useABK="1">
                <cutout left="-5.68434e-14" scale="0.253672" top="-480.503"/>
                <ClipartConfiguration applySpotColor="0"/>
            </image>
        </area>
        <area areatype="textarea">
            <position height="236.546" left="89.8876" rotation="0" top="808.989" width="553.519" zposition="7000"/>
            <decoration/>
            <text applySpotColor="0" areaTextType="content"><![CDATA[<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd"><html><head><meta name="qrichtext" content="1" /><style type="text/css">p, li { white-space: pre-wrap; }</style></head><body style=" font-family:'Century Gothic'; font-size:16pt; font-weight:400; font-style:normal;"><table style="-qt-table-type: root; margin-top:13.8434px; margin-bottom:13.8434px; margin-left:13.8434px; margin-right:13.8434px;"><tr><td style="border: none;"><p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#000000;">Gruppenfotos mit den Schultüten</span></p></td></tr></table></body></html>]]><outline color="#ff000000" width="2"/>
                <textFormat Alignment="ALIGNLEFT" IndentMargin="13.843351548269579" VerticalIndentMargin="50" backgroundColor="#00000000" font="Century Gothic,16,-1,5,50,0,0,0,0,0" foregroundColor="#ff000000" hasOutline="1" outlineColor="#ff000000" outlineWidth="2"/>
            </text>
        </area>
        <area areatype="textarea">
            <position height="150" left="2053.22" rotation="0" top="652.868" width="1523.36" zposition="7001"/>
            <decoration/>
            <text applySpotColor="0" areaTextType="content"><![CDATA[<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd"><html><head><meta name="qrichtext" content="1" /><style type="text/css">p, li { white-space: pre-wrap; }</style></head><body style=" font-family:'Century Gothic'; font-size:16pt; font-weight:400; font-style:normal;"><table style="-qt-table-type: root; margin-top:13.8434px; margin-bottom:13.8434px; margin-left:13.8434px; margin-right:13.8434px;"><tr><td style="border: none;"><p style=" margin-top:0px; margin-bottom:0px; margin-left:0px; margin-right:0px; -qt-block-indent:0; text-indent:0px;"><span style=" color:#000000;">Frische Wiese fressen Goldie und Fiona am liebsten.</span></p></td></tr></table></body></html>]]><outline color="#ff000000" width="2"/>
                <textFormat Alignment="ALIGNLEFT" IndentMargin="13.843351548269579" VerticalIndentMargin="50" backgroundColor="#00000000" font="Century Gothic,16,-1,5,50,0,0,0,0,0" foregroundColor="#ff000000" hasOutline="1" outlineColor="#ff000000" outlineWidth="2"/>
            </text>
        </area>
    </page>`

**Comment by AnEnglishmanInNorway** [comment]
I wonder if maybe the fixes for issue https://github.com/bash0/cewe2pdf/issues/56 might have had some effect on this? I see the xml has a tag for clipartconfiguration, and one of the problems I had with clipart transparency was black edges. Could you check again after you've taken in my fixes?

**Comment by steff517** [comment]
The reason for the border is the passepartout. Here's the page with and without in CEWE

With
![image](https://user-images.githubusercontent.com/1554347/102451253-6a422c80-4038-11eb-997a-58b4d3261853.png)
Without
![image](https://user-images.githubusercontent.com/1554347/102451304-81811a00-4038-11eb-8743-6feafaf17771.png)

With passepartout the images are scaled down. Since cewe2pdf does not render the passepartout, there's a black frame. 

Here's the difference with/without passepartout in code:
![image](https://user-images.githubusercontent.com/1554347/102451120-323ae980-4038-11eb-85ab-6fbf34cf39b4.png)

The ID of
`<designElementIDs passepartout="74128"/>`
can be found here:
 C:\ProgramData\hps\%KEYACCOUNT%\addons\74128

Here's the folder zipped:
[63488.zip](https://github.com/bash0/cewe2pdf/files/5707363/63488.zip)

**Comment by AnEnglishmanInNorway** [comment]
Once again you are right, Stefan, frames (passepartout) are not coded. I have just committed an update to the unittest mcf to show that, and a TODO comment in the code where I think it should be handled. But I'm not going to pick up that job now, it will involve quite a bit of investigation as to how the 2 or 3 .clp files associated with a frame should be loaded and output to the page.

**Comment by steff517** [comment]
👍

**Comment by AnEnglishmanInNorway** [comment]
@BarchSteel, I think perhaps you're the one of us that knows most about image processing. Do you see yourself having the opportunity to look at this?

**Comment by steff517** [comment]
An acceptable workaround for me would be to resize the image to fit the frame. Even if there is no passepartout, at least there's no black border.

**Comment by AnEnglishmanInNorway** [comment]
Yes, that would be a good idea, but ... here's a diff between the photo decorated with a border and shadow defined (top) and then with a passepartout instead. 
![image](https://user-images.githubusercontent.com/12747177/102629713-cba6f000-414b-11eb-801c-d06afec208db.png)
I don't immediately see how to calculate the size without the passepartout. I have a suspicion that dropping the passepartout onto the photo adjusts the displayed part of the photo based on the size of the passepartout original. There's some information in the xml file in the addon directory for the passepartout I was using, shown below. Perhaps, with a little investigation, we could figure out how to work back from the values there to an appropriate scaling of the photo??
![image](https://user-images.githubusercontent.com/12747177/102630192-87681f80-414c-11eb-90d0-d3b7ea214dcf.png)

**Comment by steff517** [comment]
Since it's a workaround, could you just go for a safe, fixed value like 15 pixels? Totally acceptable for me. If you introduce that value as a global variable, I could do some testing to figure out a good value.

**Comment by AnEnglishmanInNorway** [comment]
I don't see how a fixed value would work as a workaround - the value surely depends on how much the photo and frame have been scaled. I went back to the unittest file and made **just one change**, deleting the frame on the one picture I had framed. The album editor always seems to make some gratuitous changes, but the only change in the image description was this
![image](https://user-images.githubusercontent.com/12747177/102690487-0e2d0300-4206-11eb-8c01-8e56572779aa.png)
Apart from the missing passepartout id, there were simply new "left" and "scale" attribute values in the cutout tag. Any workaround we implemented would have to check if the image had a passepartout id and then (knowing that we do not implement passepartout) adjust the cutout attribute values left and scale (and perhaps top) by some "magically guessed" amount - an amount which should surely depend on the actual size of frame being used (which of course we don't know without starting to do a full passepartout implementation.) 
Basically I fear that any "magic guessing" algorithm would never be good enough and would ultimately cause us more trouble than doing the work to analyse the xml describing the frame and figuring out how to use it. On a quick look I certainly couldn't see how the values in the frame xml could lead to the two changes in the left and scale attributes.
I'm sorry, I just don't know where to go with this one, at least without a much bigger investment in time - and I'm already way over the time I had planned to use right now. I will say that if you can actually figure out an algorithm to use the numbers in the frame description to "remove" the frame adjustments and just leave the unframed picture in the right place and with the right size, then I'll see if I can find time to code it  - that would be a good step forward. And then later somebody might be able to figure out how to actually add the frames to the pdf.

**Comment by BarchSteel** [comment]
Even finding the right file in the general case is tricky.
For example, for my version there is the file: 
Resources\photofun\decorations\summerholiday_frames\12195-DECO-SILVER-GD\12195-DECO-SILVER-GD.xml
it contains the definition for the `<decoration designElementId="122048" ...`
However, the filename of the .xml does not contain the this designElementID! There is no other way than to build an an index of all .xml files and look into each one, which designElementID it defines.
There are some passepartouts where in the directory lies a .json with the same name as the designElementID (like the case for designElementId="125186"). Not sure why there is an extra .json for this, but not for the other. Also, the meta-data in .json and .xml do not match....

For the general approach it seems like this: The "cutout" child element in the .mcf specifies the portion of the photo to be shown.
The <position height= and width= ...> determine the bounding box for the passpartout.
In the passpartput .xml the clipping = mask = fading file is defined, and also the file for the border graphics.
The <fotoarea /> tag defines the position and size of where the clipping mask should be located.

To sum up, that's a lot of work.

**Comment by AnEnglishmanInNorway** [comment]
@BarchSteel Thanks for all the work before the vacation! Just brought in all the latest changes and done some quick testing. It seems to me that you are more organized about unit tests than the simple "look and see" strategy of using unittest_fotobook.mcf. I tried that strategy and observe that the one passepartout example I had added there isn't quite right yet. I'll attach a picture showing (left side) the passepartout example before the new changes and then (right side) after. The frame is there OK, so that's the major improvement. However, the picture itself still has the black bottom (not clipped properly?) and seems to have moved down and right. I will try to look at this myself in the course of a week or three (skiing season here!) unless you're so intrigued that you can't resist ;-)
![image](https://user-images.githubusercontent.com/12747177/103810371-729bef00-505b-11eb-9424-b890907ad3ca.png)

**Comment by fischerscode** [comment]
FYI, I had a similar problem with my blurry shadows. I let AI vibe code a good enough fix for my use case. Feel free to [give it a try](https://github.com/fischerscode/cewe2pdf/commit/4d5c2afccedd7e89d3ba0b03c0cc2ea826833c90) if you use blurred shadows.

_Since it's vibe coded, only tested against my use case and as I'm not familiar with the project, I won't create a PR._

**Comment by AnEnglishmanInNorway** [comment]
> FYI, I had a similar problem with my blurry shadows. I let AI vibe code a good enough fix for my use case. Feel free to [give it a try](https://github.com/fischerscode/cewe2pdf/commit/4d5c2afccedd7e89d3ba0b03c0cc2ea826833c90) if you use blurred shadows.
> 
> _Since it's vibe coded, only tested against my use case and as I'm not familiar with the project, I won't create a PR._

Hi @fischerscode. I have just done a lot of work on the project with the aid of the codex AI, amongst other things fixing the shadowing code. I asked it to comment on the changes you had made, here is what it said:

Yes. It is a 1-file commit titled “Refactor shadow processing and add blur functionality”: 150 lines added and 61 removed in the old monolithic cewe2pdf.py. It is very recognisably the ancestor of the shadow work we have just done:

- adds PIL.ImageFilter;
- passes the cropped image into shadow processing;
- derives a silhouette from it;
- creates and draws a blurred shadow image;
- separates parsing of angle, distance, intensity, spread/overhang, and blur.

The key difference is that it appears to use the image silhouette for every shadow, whereas our current code has been made more careful about the geometric cases, corner shapes, and the measured CEWE parameter behaviour.
So: definitely worth acknowledging in the history and perhaps crediting in comments/README if we have not already. It also confirms that the blurred-silhouette direction was not an invention from nowhere — it was a sensible approach previously explored in that fork.

So, the current base version should work for your stuff as well if you want to try it
